File: 1 - metodosBusca/python/app/core/busca/estado.py
from abc import ABC, abstractmethod
from typing import List

class Estado(ABC):
    @abstractmethod
    def get_descricao(self) -> str:
        """
        Retorna uma descrição do problema que esta representação de estado resolve.
        """
        pass

    @abstractmethod
    def eh_meta(self) -> bool:
        """
        Verifica se o estado é meta.
        """
        pass

    @abstractmethod
    def custo(self) -> int:
        """
        Custo para geração deste estado (não é o custo acumulado --- g).
        """
        pass

    @abstractmethod
    def sucessores(self) -> List['Estado']:
        """
        Gera uma lista de sucessores do nodo.
        """
        pass

File: 1 - metodosBusca/cSharp/Estado.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;


namespace cSharp
{
    /**
     * Representa um estado do mundo e as transicoes possiveis
     */
    public interface Estado
    {

        /**
         * retorna uma descricao do problema que esta representacao
         * de estado resolve
         */
        public String getDescricao();

        /**
         * verifica se o estado e meta 
         */
        public bool ehMeta();


        /**
         * Custo para geracao deste estado
         * (nao e o custo acumulado --- g)
         */
        public int custo();


        /**
         * gera uma lista de sucessores do nodo.
         */        
        public List<Estado> sucessores();
    }
}
